src/wp-includes/blocks.php

```diff
@@ -402,32 +418,22 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 	 * instead of reading a JSON file per-block, and then decoding from JSON to PHP.
 	 * Using a static variable ensures that the metadata is only read once per request.
 	 */
-	static $core_blocks_meta;
-	if ( ! $core_blocks_meta ) {
-		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
-	}
 
 	$metadata_file = ( ! str_ends_with( $file_or_folder, 'block.json' ) ) ?
 		trailingslashit( $file_or_folder ) . 'block.json' :
 		$file_or_folder;
 
-	$is_core_block = str_starts_with( $file_or_folder, ABSPATH . WPINC );
-	// If the block is not a core block, the metadata file must exist.
-	$metadata_file_exists = $is_core_block || file_exists( $metadata_file );
+	$registry_metadata = WP_Block_Metadata_Registry::get_metadata( $file_or_folder );
+	// If the block is not registered in the metadata registry, the metadata file must exist.
+	$metadata_file_exists = $registry_metadata || file_exists( $metadata_file );
 	if ( ! $metadata_file_exists && empty( $args['name'] ) ) {
 		return false;
 	}
 
-	// Try to get metadata from the static cache for core blocks.
-	$metadata = array();
-	if ( $is_core_block ) {
-		$core_block_name = str_replace( ABSPATH . WPINC . '/blocks/', '', $file_or_folder );
-		if ( ! empty( $core_blocks_meta[ $core_block_name ] ) ) {
-			$metadata = $core_blocks_meta[ $core_block_name ];
-		}
-	}
+	// Try to get metadata from the metadata registry.
+	$metadata = $registry_metadata ?? array();
 
-	// If metadata is not found in the static cache, read it from the file.
+	// If metadata is not found in the metadata registry, read it from the file.
 	if ( $metadata_file_exists && empty( $metadata ) ) {
 		$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
 	}
```

Comment:
I think we can make this code easier to follow: Right now this is an `if else` (either use the registry or read from the metadata file) disguised as something else, which makes it harder to understand.

Another note: The `$metadata_file_exists` variable should continue to rely on whether the file actually exists. We don't know that from whether the `$registry_metadata` is available. This was different for the previous `$is_core_block` condition, since for Core we could safely assume that the file is present, but for a plugin we don't know.

Something like:
```php
$metadata_file_exists = file_exists( $metadata_file );

$metadata = WP_Block_Metadata_Registry::get_metadata( $file_or_folder );
if ( ! $metadata ) {
	if ( ! $metadata_file_exists ) {
		return false;
	}
	$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
}
```

Reply:
So I think there are three main ways to use the function, via block.json file, via the registry we added, and via $args (can be useful to check cc2133fc3448 where this one was added).

I reorganized it in a way that makes sense to me. What do you think?

Ideally, I'd like to avoid the file_exists call when using this new method, but `$metadata_file_exists` does seem to be used later in the function, which I missed previously - good find. My reasoning is I'd like the plugins to be able to register their 50-100 blocks without disk access at all, but this may be acceptable since an existing `file_exists()` is cached by PHP in a way that the operations done by `wp_json_file_decode()` are not.

Reply:
I think we can still reduce some of this code, but logically your current iteration looks good to me. I think we should avoid multiple assignments in if clauses though for `$metadata_file_exists`, as it can be assigned just one - no need to try to avoid `file_exists()` checks in the logic when it'll have to happen anyway. As you're mentioning, that shouldn't really be a performance concern.